### PR TITLE
[Data2Vec] Add data2vec vision

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Finishes the Data2Vec integration by adding https://huggingface.co/models?other=data2vec-vision from https://github.com/facebookresearch/data2vec_vision/tree/main/beit

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/beit/modeling_beit.py [hunk 1]
```diff
@@ -702,7 +702,8 @@ def forward(
         pooled_output = self.pooler(sequence_output) if self.pooler is not None else None
 
         if not return_dict:
-            return (sequence_output, pooled_output) + encoder_outputs[1:]
+            head_outputs = (sequence_output, pooled_output) if pooled_output is not None else (sequence_output,)
```
Comment (patrickvonplaten) [inline src/transformers/models/beit/modeling_beit.py:705]: @NielsRogge we should not pass `None` values here as this would not pass an equivalence check in case the pooling layer is not added

File: src/transformers/models/beit/modeling_beit.py [hunk 2]
```diff
@@ -713,7 +714,7 @@ def forward(
 
 
 class BeitPooler(nn.Module):
-    def __init__(self, config: BeitModel) -> None:
+    def __init__(self, config: BeitConfig) -> None:
```
Comment (patrickvonplaten) [inline src/transformers/models/beit/modeling_beit.py:717]: typo

File: src/transformers/models/beit/modeling_beit.py [hunk 3]
```diff
@@ -817,7 +818,7 @@ def forward(
             masked_lm_loss = loss_fct(prediction_scores[bool_masked_pos], labels)
 
         if not return_dict:
-            output = (prediction_scores,) + outputs[2:]
+            output = (prediction_scores,) + outputs[1:]
```
Comment (patrickvonplaten) [inline src/transformers/models/beit/modeling_beit.py:821]: Adapted since there is no `None` value anymore now

File: src/transformers/models/data2vec/configuration_data2vec_vision.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
+# coding=utf-8
+# Copyright Microsoft Research and The HuggingFace Inc. team. All rights reserved.
```
Comment (patrickvonplaten) [inline src/transformers/models/data2vec/configuration_data2vec_vision.py:2]:
```suggestion
# Copyright Facebook and The HuggingFace Inc. team. All rights reserved.
```

**Comment by patrickvonplaten** [comment]
Uploaded checkpoints are here: https://huggingface.co/models?other=data2vec-vision . Will add a README to those ones and all other data2vec ones after this PR is merged

File: src/transformers/models/data2vec/modeling_data2vec_vision.py [hunk 5]
```diff
@@ -0,0 +1,1209 @@
+# coding=utf-8
+# Copyright 2021 Microsoft Research and The HuggingFace Inc. team. All rights reserved.
```
Comment (sgugger) [inline src/transformers/models/data2vec/modeling_data2vec_vision.py:2]:
```suggestion
# Copyright 2022 Microsoft Research and The HuggingFace Inc. team. All rights reserved.
```

**Review by sgugger — APPROVED** [review]
Thanks for adding this new model!

File: docs/source/en/model_doc/data2vec.mdx [hunk 6]
```diff
@@ -33,10 +33,13 @@ Models and code are available at www.github.com/pytorch/fairseq/tree/master/exam
 
 Tips:
 
-- Both Data2VecAudio and Data2VecText have been trained using the same self-supervised learning method.
-  In the case of Data2VecAudio, preprocessing is identical to [`RobertaModel`], including tokenization.
+- Both Data2VecAudio, Data2VecText, and Data2VecVision have been trained using the same self-supervised learning method.
```
Comment (NielsRogge) [inline docs/source/en/model_doc/data2vec.mdx:36]:
```suggestion
- Data2VecAudio, Data2VecText, and Data2VecVision have all been trained using the same self-supervised learning method.
```

File: src/transformers/models/beit/modeling_beit.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/beit/modeling_beit.py:705]: Thanks for fixing!

File: src/transformers/models/data2vec/configuration_data2vec_vision.py (hunk 4 above)
Comment (NielsRogge) [inline src/transformers/models/data2vec/configuration_data2vec_vision.py:2]:
```suggestion
# Copyright Meta Platforms and The HuggingFace Inc. team. All rights reserved.
```

File: src/transformers/models/data2vec/configuration_data2vec_vision.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+
+class Data2VecVisionConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`Data2VecVisionModel`]. It is used to instantiate
+    an Data2VecVision model according to the specified arguments, defining the model architecture. Instantiating a
+    configuration with the defaults will yield a similar configuration to that of the Data2VecVision
+    [microsoft/data2vec_vision-base-patch16-224-in22k](https://huggingface.co/microsoft/data2vec_vision-base-patch16-224-in22k)
+    architecture.
```
Comment (NielsRogge) [inline src/transformers/models/data2vec/configuration_data2vec_vision.py:39]: Should this become Facebook?

File: src/transformers/models/data2vec/configuration_data2vec_vision.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            Number of convolutional layers to use in the auxiliary head.
+        auxiliary_concat_input (`bool`, *optional*, defaults to `False`):
+            Whether to concatenate the output of the auxiliary head with the input before the classification layer.
+        semantic_loss_ignore_index (`int`, *optional*, defaults to 255):
+            The index that is ignored by the loss function of the semantic segmentation model.
+
+    Example:
+
+    ```python
+    >>> from transformers import Data2VecVisionModel, Data2VecVisionConfig
```
Comment (NielsRogge) [inline src/transformers/models/data2vec/configuration_data2vec_vision.py:105]:
```suggestion
    >>> from transformers import Data2VecVisionConfig, Data2VecVisionModel
```

File: src/transformers/models/data2vec/modeling_data2vec_vision.py (hunk 5 above)
Comment (NielsRogge) [reply src/transformers/models/data2vec/modeling_data2vec_vision.py:2]: Wondering if we should cite both Microsoft (for BEiT) and Facebook (for Data2Vec)

File: src/transformers/models/data2vec/modeling_data2vec_vision.py [hunk 9]
```diff
@@ -0,0 +1,1209 @@
...
+        if self.position_embeddings is not None:
+            embeddings = embeddings + self.position_embeddings
+        embeddings = self.dropout(embeddings)
+
+        return embeddings
+
+
+# Based on timm implementation, which can be found here:
+# https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py
+class PatchEmbeddings(nn.Module):
```
Comment (NielsRogge) [inline src/transformers/models/data2vec/modeling_data2vec_vision.py:190]: Any reason there's no copied from statement here?

File: src/transformers/models/data2vec/modeling_data2vec_vision.py [hunk 10]
```diff
@@ -0,0 +1,1209 @@
...
+        return drop_path(x, self.drop_prob, self.training)
+
+    def extra_repr(self) -> str:
+        return "p={}".format(self.drop_prob)
+
+
+# Based on timm implementation, which can be found here:
+# https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py
+# Copied from transformers.models.beit.modeling_beit.BeitEmbeddings with Beit->Data2VecVision
+class Data2VecVisionEmbeddings(nn.Module):
```
Comment (NielsRogge) [inline src/transformers/models/data2vec/modeling_data2vec_vision.py:141]: Can we include a copied from statement here?

File: src/transformers/models/data2vec/modeling_data2vec_vision.py [hunk 11]
```diff
@@ -0,0 +1,1209 @@
...
+        return x
+    keep_prob = 1 - drop_prob
+    shape = (x.shape[0],) + (1,) * (x.ndim - 1)  # work with diff dim tensors, not just 2D ConvNets
+    random_tensor = keep_prob + torch.rand(shape, dtype=x.dtype, device=x.device)
+    random_tensor.floor_()  # binarize
+    output = x.div(keep_prob) * random_tensor
+    return output
+
+
+class DropPath(nn.Module):
```
Comment (NielsRogge) [inline src/transformers/models/data2vec/modeling_data2vec_vision.py:124]: Copied from

File: src/transformers/models/data2vec/modeling_data2vec_vision.py [hunk 12]
```diff
@@ -0,0 +1,1209 @@
...
+# https://github.com/rwightman/pytorch-image-models/blob/b9bd960a032c75ca6b808ddeed76bee5f3ed4972/timm/models/layers/helpers.py
+# From PyTorch internals
+def to_2tuple(x):
+    if isinstance(x, collections.abc.Iterable):
+        return x
+    return (x, x)
+
+
+# Based on https://github.com/rwightman/pytorch-image-models/blob/a2727c1bf78ba0d7b5727f5f95e37fb7f8866b1f/timm/models/layers/drop.py
+def drop_path(x: torch.Tensor, drop_prob: float = 0.0, training: bool = False) -> torch.Tensor:
```
Comment (NielsRogge) [inline src/transformers/models/data2vec/modeling_data2vec_vision.py:104]: Copied from

File: tests/data2vec/test_modeling_data2vec_vision.py [hunk 13]
```diff
@@ -0,0 +1,450 @@
+# coding=utf-8
+# Copyright 2021 The HuggingFace Inc. team. All rights reserved.
```
Comment (NielsRogge) [inline tests/data2vec/test_modeling_data2vec_vision.py:2]:
```suggestion
# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```

**Review by NielsRogge — APPROVED** [review]
Thanks for adding! My only comment is about some additional copied from statements, if possible.

File: tests/data2vec/test_modeling_data2vec_vision.py [hunk 14]
```diff
@@ -0,0 +1,450 @@
...
+        for model_class in self.all_model_classes:
+            if model_class in [*get_values(MODEL_MAPPING)]:
+                continue
+            # TODO: remove the following 3 lines once we have a MODEL_FOR_SEMANTIC_SEGMENTATION_MAPPING
+            # this can then be incorporated into _prepare_for_class in test_modeling_common.py
+            elif model_class.__name__ == "Data2VecVisionForSemanticSegmentation":
+                batch_size, num_channels, height, width = inputs_dict["pixel_values"].shape
+                inputs_dict["labels"] = torch.zeros(
+                    [self.model_tester.batch_size, height, width], device=torch_device
+                ).long()
```
Comment (NielsRogge) [inline tests/data2vec/test_modeling_data2vec_vision.py:243]: These lines can be removed, since labels are now created in `test_modeling_common.py`

File: src/transformers/models/data2vec/modeling_data2vec_vision.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/transformers/models/data2vec/modeling_data2vec_vision.py:2]: Thank just Meta Platforms is good here since they forked BEIT

File: src/transformers/models/data2vec/modeling_data2vec_vision.py (hunk 9 above)
Comment (patrickvonplaten) [reply src/transformers/models/data2vec/modeling_data2vec_vision.py:190]: Good point!

File: src/transformers/models/data2vec/modeling_data2vec_vision.py (hunk 10 above)
Comment (patrickvonplaten) [reply src/transformers/models/data2vec/modeling_data2vec_vision.py:141]: Think it's there already ;-)

File: src/transformers/models/data2vec/configuration_data2vec_vision.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+    "facebook/data2vec-vision-base-ft": "https://huggingface.co/facebook/data2vec-vision-base-ft/resolve/main/config.json",
+}
+
+
+class Data2VecVisionConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`Data2VecVisionModel`]. It is used to instantiate
+    an Data2VecVision model according to the specified arguments, defining the model architecture. Instantiating a
+    configuration with the defaults will yield a similar configuration to that of the Data2VecVision
+    [microsoft/data2vec_vision-base-patch16-224-in22k](https://huggingface.co/microsoft/data2vec_vision-base-patch16-224-in22k)
```
Comment (patrickvonplaten) [inline src/transformers/models/data2vec/configuration_data2vec_vision.py:38]:
```suggestion
    [facebook/data2vec-vision-base](https://huggingface.co/facebook/data2vec-vision-base)
```
